wip-trampoline.md

```diff
@@ -0,0 +1,780 @@
+  * [Merchant without trampoline support](#merchant-without-trampoline-support)
+* [Appendix B: Trampoline and Rendez-Vous](#appendix-b-trampoline-and-rendez-vous)
+* [Appendix C: Trampoline and MPP](#appendix-c-trampoline-and-mpp)
+* [References](#references)
+
+## Features
+
+Trampoline routing uses the following `globalfeatures` flags:
+
+| Bits  | Name                 | Description                                    | Link |
+|-------|----------------------|------------------------------------------------|------|
+| 10/11 | `trampoline_routing` | This node requires/supports trampoline routing | TBD  |
+
+Feature flags are assigned in pairs, but the mandatory value may not be applicable here.
+
+Gossip messages filtering uses the following `localfeatures` flags:
+
+| Bits  | Name                 | Description                                        | Link |
+|-------|----------------------|----------------------------------------------------|------|
+| 12/13 | `gossip_filters`     | This node requires/supports gossip message filters | TBD  |
+
+## Packet Structure
+
+### Trampoline Onion
+
+The trampoline onion is a fixed-size tlv packet with the following structure:
+
+1. type: 14 (`trampoline_onion_packet`)
+2. data:
+   * [`byte`:`version`]
+   * [`point`:`public_key`]
+   * [`400*byte`:`hop_payloads`]
+   * [`32*byte`:`hmac`]
+
+This results in a `466`-bytes fixed-size tlv where the value has exactly the same format as the `onion_packet` with
+a smaller `hop_payloads` (`400` bytes instead of `1300`).
+
+Trampoline `hop_payload`s may contain the following fields:
+
+1. tlvs: `trampoline_tlv_payload`
+2. types:
+    * type: 2 (`amt_to_forward`)
+    * data:
+        * [`tu64`:`amt_to_forward`]
+    * type: 4 (`outgoing_cltv_value`)
+    * data:
+        * [`tu32`:`outgoing_cltv_value`]
+    * type: 10 (`next_node_id`)
+    * data:
+        * [`point`:`next_node_id`]
+    * type: 12 (`recipient_features`)
+    * data:
+        * [`length*byte`:`recipient_features`]
+
+### Paying via trampoline nodes
+
+A node that wants to rely on trampoline nodes to relay payments should use a `trampoline_onion_packet` in the
+`hop_payload` of the _last_ hop of a normal `onion_packet`:
+
+1. type: `onion_packet`
+2. data:
+   * [`byte`:`version`]
+   * [`point`:`public_key`]
+   * [`n1*byte`:`hop_payload`] (normal hop payload)
+   * [`n2*byte`:`hop_payload`] (normal hop payload)
+   * ...
+   * [`nn*byte`:`hop_payload`] (hop payload containing a `trampoline_onion_packet`)
+   * `filler`
+   * [`32*byte`:`hmac`]
+
+### Paying to non-trampoline nodes
+
+If the recipient of the payment doesn't support trampoline routing, the last trampoline node must convert the last hop
+to a standard onion payment. The `recipient_features` tlv is used to detect that the next node doesn't support
+trampoline.
+
+Note that it reveals the identity of the recipient and the amount paid to the last trampoline node (but it doesn't
```

Comment:
Payer can always leak payee privacy, even today (since payer has this information). It can put information in ignored bits at the end of any message, or in an odd tlv field, etc.
I strongly encourage payees to support trampoline because it protects their privacy and automatically re-balances channel capacity (by routing payments).
But in order to have a reasonable upgrade path, we need to support paying to legacy nodes that don't know about trampoline (which is why this feature is necessary). I'm not happy about it, but I don't see a better solution at the moment...

Reply:
I think there is a difference between intentionally having to tweak protocol messages to leak info and just leaking info by following faithful execution of protocol. Not happy too, I'm wandering if we can solve it with another routing node of indirection but naive solutions I can think of are gross..

Reply:
> I think there is a difference between intentionally having to tweak protocol messages to leak info and just leaking info by following faithful execution of protocol

That's true. There is definitely room for improvement here, I'll think about this further too. If you find a work-around let me know ;)

Reply:
I'm leaning more and more towards removing the ability to pay legacy nodes via this trampoline-to-legacy mechanism (unless we find a better way to do it than what I currently have).